usvm-ts-dataflow/src/main/kotlin/org/usvm/dataflow/ts/ifds/EtsForwardSourceRunner.kt

```diff
@@ -0,0 +1,130 @@
+package org.usvm.dataflow.ts.ifds
+
+import org.jacodb.ets.model.EtsStmt
+import org.usvm.dataflow.ifds.Edge
+import org.usvm.dataflow.ifds.Vertex
+import org.usvm.dataflow.ts.ifds.EtsForwardMethodRunner.PathEdge
+import org.usvm.dataflow.ts.util.EtsTraits
+
+class EtsForwardSourceRunner<Fact>(
+    val traits: EtsTraits,
+    val methodRunner: EtsForwardMethodRunner<Fact, *>,
+    val startingFact: Fact,
+) {
+    private var enqueued: Boolean = false
+    private val internalQueue: ArrayDeque<PathEdge<Fact>> = ArrayDeque()
+
+    fun processFacts() {
+        while (!internalQueue.isEmpty()) {
+            val currentEdge = internalQueue.removeFirst()
+            tabulationAlgorithmStep(currentEdge)
+        }
+        enqueued = false
+    }
+
+    private val cfg = methodRunner.cfg
+    private val successors = methodRunner.successors
+    private val stmts = methodRunner.stmts
+    private val mockStmt = methodRunner.mockStmt
+    private val entrypoint = methodRunner.entrypoint
+
+    private val callToReturnSiteFlowFunction = methodRunner.callToReturnSiteFlowFunction
+    private val callToStartFlowFunctions = methodRunner.callToStartFlowFunctions
+    private val sequentFlowFunction = methodRunner.sequentFlowFunction
+
+    private val factsAtStmt = Array(stmts.size) { hashSetOf<Fact>() }
+
+    internal fun propagate(edge: PathEdge<Fact>) {
+        val (ip, fact) = edge
+        if (factsAtStmt[ip].add(fact)) {
+            val startVertex = Vertex(mockStmt, startingFact)
+            val endVertex = Vertex(stmts[ip], fact)
+            for (event in methodRunner.analyzer.handleNewEdge(Edge(startVertex, endVertex))) {
+                methodRunner.manager.handleEvent(event)
+            }
+
+            internalQueue.add(edge)
+            if (!enqueued) {
+                enqueued = true
+                methodRunner.enqueue(this)
+            }
+        }
+    }
+
+    private val callerPathEdges = hashSetOf<Edge<Fact, EtsStmt>>()
+    internal val summaryEdges = hashSetOf<Vertex<Fact, EtsStmt>>()
+
+    private fun tabulationAlgorithmStep(
+        currentEdge: PathEdge<Fact>,
+    ) = with(traits) {
+        val (currentIp, currentFact) = currentEdge
+        val current = stmts[currentIp]
+
+        val currentIsCall = getCallExpr(current) != null
+        val currentIsExit = current in cfg.exits
```

Comment:
This check (`.exits` might be expensive to compute) is unnecessary if the current statement is a call.